### PR TITLE
Add vectorized batch_norm_cpu_collect_stats_contiguous_impl

### PR DESCRIPTION
**Pull request by pratiklp00**
This patch Adds the  vectorized  implementation of batch_norm_cpu_collect_stats_contiguous_impl.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by janeyx99** [comment]
@jgong5 assigning you as reviewer, please reassign if you are not the best reviewer for this PR.

File: aten/src/ATen/native/cpu/batch_norm_kernel.cpp [hunk 1]
```diff
@@ -192,27 +189,44 @@ batch_norm_cpu_collect_stats_contiguous_impl(
   // parallel dim reduce on 'channel'
   at::parallel_for(0, n_channel, 1, [&](int64_t begin, int64_t end) {
     for (const auto c : c10::irange(begin, end)) {
-      // compute mean per input
-      accscalar_t sum = 0;
-      for (const auto n : c10::irange(n_batch)) {
-        for (const auto i : c10::irange(image_size)) {
-          auto offset = n * n_channel * image_size + c * image_size + i;
-          sum += input_data[offset];
+    // compute mean per input
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/batch_norm_kernel.cpp:192]: nit: indent

**Review by jgong5 — APPROVED** [review]

File: aten/src/ATen/native/cpu/batch_norm_kernel.cpp [hunk 2]
```diff
@@ -192,27 +189,44 @@ batch_norm_cpu_collect_stats_contiguous_impl(
...
+        for (; d < image_size - (image_size % Vec::size()); d += Vec::size()) {
+          Vec data_vec = Vec::loadu(input_ptr + d);
+          sum_val_vec += data_vec;
+        }
+        for (; d < image_size; d++) {
+          sum_val += input_ptr[d];
         }
       }
-      scalar_t mean = sum / N;
+      sum_val += vec_reduce_all([](Vec& x, Vec& y) { return x + y; }, sum_val_vec);
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/batch_norm_kernel.cpp:206]: `vec_reduce_all(const Op&, const Vectorized<scalar_t>&)` here is already the `fast` version: https://github.com/pytorch/pytorch/tree/main/aten/src/ATen/cpu#L82C17-L82C31, it goes into a vectorized horizontal reduce of SIMD

`vec_reduce_all(const Op&, const Vectorized<scalar_t>&), int64_t` is a slow version when the size < vector length.

File: aten/src/ATen/native/cpu/batch_norm_kernel.cpp [hunk 3]
```diff
@@ -192,27 +189,44 @@ batch_norm_cpu_collect_stats_contiguous_impl(
...
     for (const auto c : c10::irange(begin, end)) {
-      // compute mean per input
-      accscalar_t sum = 0;
-      for (const auto n : c10::irange(n_batch)) {
-        for (const auto i : c10::irange(image_size)) {
-          auto offset = n * n_channel * image_size + c * image_size + i;
-          sum += input_data[offset];
+    // compute mean per input
+      scalar_t sum_val = 0;
+      Vec sum_val_vec = Vec(scalar_t(0));
```
Comment (mingfeima) [inline aten/src/ATen/native/cpu/batch_norm_kernel.cpp:194]: you can use a simple `vec::map_reduce` so save some effort: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/cpu/vec/functional_base.h#L137

**Review by mingfeima — COMMENTED** [review]
As long as CI passes, I am fine with it. Otherwise you have to deal with the numerical stability issue for large feature map plane.

**Review by malfet — CHANGES_REQUESTED** [review]
If goal of this PR is to increase performance I think it needs some before and after performance numbers, otherwise it does not make much sense.

Also, considering that you are chaning accumulation type from float to double would be interested to see some analysis on accuracy of old vs new op implementation

File: aten/src/ATen/native/cpu/batch_norm_kernel.cpp [hunk 4]
```diff
@@ -176,10 +176,7 @@ template <typename scalar_t>
 typename std::enable_if_t<std::is_same_v<scalar_t, at::opmath_type<scalar_t>>, void>
 batch_norm_cpu_collect_stats_contiguous_impl(
     Tensor& mean, Tensor& var_sum, const Tensor& input) {
-
```
[the remark this answers is not in the document]
Comment (pratiklp00) [reply aten/src/ATen/native/cpu/batch_norm_kernel.cpp:179]: Hi @mingfeima as you said after the changes 3 test cases are failing from file test_deferred_batch_norm.py. if you have any suggestions please let me know.
